services/device-registry-base/src/main/java/org/eclipse/hono/service/management/BaseDto.java

```diff
@@ -0,0 +1,208 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.management;
+
+import java.time.Instant;
+import java.util.function.Supplier;
+
+import org.eclipse.hono.annotation.HonoTimestamp;
+import org.eclipse.hono.util.RegistryManagementConstants;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * The base class for implementing a DTO (Data Transfer Object) to store data in registry implementations.
+ *
+ * @param <T> The type of the data object
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public abstract class BaseDto<T> {
+
+    public static final String FIELD_UPDATED_ON = "updatedOn";
+    public static final String FIELD_VERSION = "version";
+    public static final String FIELD_DATA = "data";
+
+    @JsonProperty(value = FIELD_VERSION, required = true)
+    private String version;
+
+    @JsonProperty(RegistryManagementConstants.FIELD_STATUS_CREATION_DATE)
+    @JsonInclude(value = JsonInclude.Include.NON_EMPTY)
+    @HonoTimestamp
+    private Instant creationTime;
+
+    @JsonProperty(value = FIELD_UPDATED_ON, required = true)
+    @HonoTimestamp
+    private Instant updatedOn;
+
+    private T data;
+
+    /**
+     * Default constructor for serialisation/deserialization.
+     */
+    public BaseDto() {
+        // Explicit default constructor.
+    }
+
+    /**
+     * Constructs a new DTO for use with the <b>creation of a new</b> persistent entry.
+     *
+     * @param supplier A DTO subclass' constructor of which a new instance shall be created.
+     * @param data The data of the DTO.
+     * @param version The version of the DTO
+     *
+     * @param <P> The type of the DTO's payload.
+     * @param <T> The type of the DTO subclass.
+     *
+     * @return A DTO instance for creating a new entry.
+     */
+    public static <P, T extends BaseDto<P>> T forCreation(final Supplier<T> supplier, final P data, final String version) {
```

Comment:
In the end the idea was to encapsulate the timestamp setting for subclasses: `forCreation()` makes sure that the created timestamp ist set while `forUpdate()` takes care of the updated timestamp. Since both methods take the same arguments I think this can't be implemented using a constructor?

Reply:
why not? simply use two parameters in the constructor, one for the createdOn timestamp and another for the updatedOn timestamp, or wouldn't that work?

Reply:
Sure! :-) But, what I meant was: then it's the subclass' responsibility again to properly set the timestamps as they will pass both parameters to the constructor.

Reply:
I created a draft PR to illustrate what I mean (and to verify that I got you right): https://github.com/eclipse/hono/pull/2243/commits/9120c2e08f8a0d385604dfc3e0baa585b8cd6e35

Personally I do not see why having the constructor is better, besides that the implementation of the factory methods in BaseDto is slightly more complex as it uses generics. The downside is though that now setting the proper created and updated is now done in the subclasses meaning it is duplicated. Also the FileBasedDto cannot reuse the static factory methods of the base class.